src/utilities/include/utils.h

```diff
@@ -106,11 +106,14 @@ thrust::counting_iterator<T> countAt(T i) {
 template <typename T>
 __host__ __device__ T AtomicAdd(T& target, T add) {
 #ifdef __CUDA_ARCH__
+  // required for synchronization
+  __threadfence();
```

Comment:
Okay, so the idea of putting the fence in our `AtomicAdd` function directly is for ease of use such that they behave more consistently between CPU and GPU, right? That seems reasonable; are there any other atomic functions that should also have it? Have you noticed any performance change from this?

Reply:
Yes, this is to make them behave a bit more consistently. No, I did not notice any significant performance change from this.